### PR TITLE
Use ritelinked to replace linked-hash-map & linked_hash_set

### PR DESCRIPTION
**Pull request by PsiACE**
This PR uses ritelinked to replace linked-hash-map and linked_hash_set. In particular, ritelinked is under active maintenance and provides better performance.

Once CI is passed, it can be merged very easily .

#8883

### CODE REVIEW

File: client/db/src/lib.rs [hunk 1]
```diff
@@ -462,7 +462,15 @@ impl<Block: BlockT> sc_client_api::blockchain::HeaderBackend<Block> for Blockcha
...
 				let mut cache = self.header_cache.lock();
-				if let Some(result) = cache.get_refresh(h) {
+				if let Some(result) = {
+				        match cache.raw_entry_mut().from_key(h) {
+                                                linked_hash_map::RawEntryMut::Occupied(mut occupied) => {
+                                                        occupied.to_back();
+                                                        Some(occupied.into_mut())
+                                                }
+                                                linked_hash_map::RawEntryMut::Vacant(_) => None,
+                                       }
```
Comment (bkchr) [inline client/db/src/lib.rs:472]: We use tabs. Please fix here and below.

Comment (PsiACE) [reply client/db/src/lib.rs]: @bkchr Thanks, done.

**Comment by PsiACE** [comment]
I don't know why  continuous-integration/gitlab-test-linux-stable fails. However, this does not affect the current code is correct.

**Comment by PsiACE** [comment]
> User @PsiACE, please sign the CLA [here](https://cla.parity.io/sign_cla/36896360/3).

Done

**Comment by expenses** [comment]
IMO it'd be better to use something like https://crates.io/crates/hashlink (the crate ritelinked is forked off of) if we care about the fact that linked-hash-map/set are outdated.

**Comment by bkchr** [comment]
The crate you forked from, is still actively maintained. So, going to close this.

**Comment by PsiACE** [comment]
> IMO it'd be better to use something like https://crates.io/crates/hashlink (the crate ritelinked is forked off of) if we care about the fact that linked-hash-map/set are outdated.

- As the maintainer of **ritelinked**, I will definitely recommend my work first.
- On the other hand, **ritelinked** provides some useful features that **hashlink** does not have. I did make some improvements.
- More often, people use **hashlink** for `LruCache`, but **ritelinked** will only focus on `LinkedHashMap` and `LinkedHashSet` .

**Comment by PsiACE** [comment]
Of course, I respect your decision and keep my own opinion.

**Comment by expenses** [comment]
>     * I did make some improvements.

Which of the improvements would be relevant here?

**Comment by PsiACE** [comment]
> Which of the improvements would be relevant here?

@expenses At the moment:

- Better (de)serialization support, earlier than hashlink 0.7
- `no_std` friendly, hashlink is not at present
- the `amortized` feature, taking into account the possible long tail effect

Plans:

- (WIP) Provides better compatibility with linked-hash-map

**Comment by expenses** [comment]
We don't use any of those features.

**Comment by PsiACE** [comment]
Benchmarking ritelinked: Collecting 100 samples in estimated 5.0007 s (20M itera                                                                                ritelinked              time:   [166.25 ns 166.85 ns 167.54 ns]
Found 6 outliers among 100 measurements (6.00%)
  1 (1.00%) high mild
  5 (5.00%) high severe

Benchmarking hashlink: Collecting 100 samples in estimated 5.0010 s (20M iterati                                                                                hashlink                time:   [168.48 ns 169.34 ns 170.36 ns]
Found 13 outliers among 100 measurements (13.00%)
  3 (3.00%) high mild
  10 (10.00%) high severe

Benchmarking linked-hash-map: Collecting 100 samples in estimated 5.0010 s (17M                                                                                 linked-hash-map         time:   [365.00 ns 376.16 ns 390.56 ns]
Found 7 outliers among 100 measurements (7.00%)
  7 (7.00%) high mild

**Pull request closed without merging**